### DISCUSSION TITLE
Random questions

### DISCUSSION THREAD

**Discussion by tgkolda** [sandialabs/pyttb#345] [General]
**What is the use case for shallow copies?** I think I understand the shallow versus deep copy reasoning, but I'd still like to ask. FWIW, my impression is that MATLAB doesn't automatically do a deep copy but rather does it if it's actually needed. In any case, can you give me a use case where a deep copy would cause problems? Thanks.

**Comment by tgkolda** [comment]
**Why is elemfun its own function rather than part of the tensor class?**

**Reply by ntjohnson1** [reply]
It is a [method](https://github.com/sandialabs/pyttb/blob/70fc7c7f3d697950ed5b27fcb7f91cb5df3c6f38/pyttb/sptensor.py#L588) in the sptensor class.  I'm not sure I understand the question.

**Reply by tgkolda** [reply]
Sorry, I meant `tenfun`. Why is it `tt_tenfun` rather than a method directly in the `tensor` class?

**Reply by ntjohnson1** [reply]
This is not a deeply considered decision (open to opinions) and it isn't captured in our git history so any discussion has been lost to time.

Since it doesn't modify things in place it looks like a pure function that operates on tensors. It's usage in pyttb only really deals with the binary case of source tensor and some other tensor, but the full capabilities in matlab is more flexible than that. Where elemfun modifies things in place if I remember correctly.
```
%   Examples
%   Z = tenfun(@(x)(x+1),X) %<-- increase every element by one
%   Z = tenfun(@eq,X,1) %<-- logical comparison of X with scalar
%   Z = tenfun(@plus,X,Y) %<-- adds the two tensors X and Y.
%   Z = tenfun(@max,X,Y,Z) %<-- elementwise max over all elements in X,Y,Z
```

**Reply by tgkolda** [reply]
In MATLAB, neither operation modifies in place. So, it would be good to be consistent on that front if possible. I'm actually confused by the elemfun documentation. Does the call `S2 = S1.elemfun(lambda values: values * 2)` modify S1 or not?

**Reply by ntjohnson1** [reply]
Hmm you're right. I remembered incorrectly. Since neither modified in place I'm fine moving it back to a class method.

**Comment by tgkolda** [comment]
**Where are the various utilities documented, like tt_ind2sub?**

**Reply by ntjohnson1** [reply]
I think the answer is nowhere for some of the utilities. It looks like when we did our documentation and examples overhaul we mostly hit the more publicly facing interfaces and missed the more internally facing utilities. Opened an [issue ](https://github.com/sandialabs/pyttb/issues/347) for that oversight.

**Reply by ntjohnson1** [reply]
I made a [pass](https://github.com/sandialabs/pyttb/pull/353) over our pyttb utils to improve the documentation. Please open an issue if there are methods with lacking documentation. If you were asking if they get built into our web hosted documentation they aren't currently. On command line etc the help function brings them up. I'll add them into our web hosted docs now as a part of the Coming from MATLAB doc improvements [PR](https://github.com/sandialabs/pyttb/pull/352).

**Comment by tgkolda** [comment]
**Is it F order?**

```
import pyttb as ttb
import numpy as np
X12 = ttb.tensor(np.array([3, 9, 1, 8, 2, 1, 4, 3, 9, 6, 9, 5, 5, 6, 4, 1, 4, 1]),(3,3,2))
X12.data.flags
```

The output is:
```
  C_CONTIGUOUS : True
  F_CONTIGUOUS : False
  OWNDATA : True
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False
```

**Reply by ntjohnson1** [reply]
Numpy is C order by default. We've implemented all the indexing and algorithms to be F order for the critical pieces and do some inefficient casting. We should make everything explicitly F order, and potentially in the future support both. Delayed response because I was hoping to have time to just sit down and push through the F order plumbing but haven't had a chance.

**Reply by ntjohnson1** [reply]
This should be resolved now (at least for the case presented).

**Comment by ntjohnson1** [comment]
> What is the use case for shallow copies? I think I understand the shallow versus deep copy reasoning, but I'd still like to ask. FWIW, my impression is that MATLAB doesn't automatically do a deep copy but rather does it if it's actually needed. In any case, can you give me a use case where a deep copy would cause problems? Thanks.

I think globally doing deep copies would only really be an issue around performance (and maybe mental model if people assume they are operating on a reference rather than a copy). Do you have a particular usage that's been confusing?

**Reply by tgkolda** [reply]
Is there any place that a shallow copy is used and not a deep copy? Generally, we don't copy for no reason. I guess one question is how passing an argument into a function works. Is that a pointer or a copy?

**Reply by ntjohnson1** [reply]
Basically every mutable data structure in Python ends up being by reference. So if a deep copy is needed it must be explicit. Arguments to a function are references, but inside the function we typically add an explicit copy call if we intend to mutate that data. The numpy for matlab users doc has some more discussion on the copy/reference details.

EDIT: Original example was bad because I performed a mutation that didn't operate in place.

```python
>>> t0 = ttb.tenones((2,2,2))
# Just reference the tensor
>>> t1 = t0
>>> t1[0,0,0] = 2.0
>>> t1.isequal(t0)
True
# Confirm is a reference
>>> t1 is t0
True
# Copy of the tensor
>>> t2 = t1.copy()
>>> t2[0,0,0] = 3.0
>>> t2.isequal(t1)
False
>>>t2 is t1
False
```

**Reply by dmdunla** [reply]
> > What is the use case for shallow copies? I think I understand the shallow versus deep copy reasoning, but I'd still like to ask. FWIW, my impression is that MATLAB doesn't automatically do a deep copy but rather does it if it's actually needed. In any case, can you give me a use case where a deep copy would cause problems? Thanks.
> 
> I think globally doing deep copies would only really be an issue around performance (and maybe mental model if people assume they are operating on a reference rather than a copy). Do you have a particular usage that's been confusing?

The main use case if for pyttb integration into other Python packages that will manage their own data. Specifically, this is required when using `pygenten` (wrapper for the Genten C++ backend), as some of the data instances in `pygenten` (and really Genten deep down) create the data. We have tried to follow the conventions of Python and specifically `numpy` data classes to this end.

**Comment by tgkolda** [comment]
Ok. I would suggest that this be listed as one of the MAJOR differences between MATLAB and Python, **up at the top**. Right now, the general tensor object differences are listed as major differences (https://pyttb.readthedocs.io/en/latest/matlab/common.html#data-members). I think a bit of reorganization would be useful here. For methods that have changed more or less across all classes (like shape), it's reasonable to list them here. Otherwise, recommend putting it into a class-specific difference page.

**Reply by ntjohnson1** [reply]
I moved the copying/reference note to the top of the page.

Most of the tensors have fairly similar methods, so from a maintenance perspective keeping it in one place is nice. Otherwise we'd copy paste that table to most of the downstream pages.
